### ISSUE TITLE
Wrong execution order for `while (promise) { await promise; }`

### ISSUE THREAD

**Issue by igisev** [rpetrich/babel-plugin-transform-async-to-promises#6]
Hi
The code below demonstrates the wrong order of code execution when
`while (promise) { await promise; }`
is used.

```javascript
let index = 0;
let promise = null;
let messages = [];

async function test() {
    let promiseResolve;
    let num = ++index;

    messages.push(`start ${num}`);

    // place of interest
    while (promise) {
        messages.push(`wait ${num}`);

        await promise;
    }

    promise = new Promise(r => {
        promiseResolve = r;
    });

    await wait();

    promise = null;

    promiseResolve();

    messages.push(`stop ${num}`);
}

function wait() {
    return new Promise(resolve => setTimeout(resolve, 10));
}

Promise.all([test(), test(), test()])
    .then(() => {
        // v0.5.6 babel-plugin-transform-async-to-promises
        //             ['start 1', 'start 2', 'wait 2', 'start 3', 'wait 3', 'stop 1', 'stop 2', 'stop 3']
        let expected = ['start 1', 'start 2', 'wait 2', 'start 3', 'wait 3', 'stop 1', 'wait 3', 'stop 2', 'stop 3'];

        // false
        let isEqual = JSON.stringify(expected) === JSON.stringify(messages);

        console.log(isEqual, messages);
    });
```

**Comment by ryan-petrich-mpf** [comment]
This is a really phenomenal test case that exposed a flaw in my event loop strategy and reasoning. Thanks for contributing. This and similar cases are now fixed in 12debcd.

**Comment by rpetrich** [comment]
@igisev If you have more complex cases, it would be very helpful to have them added to the test suite.
